crunchy-bridge-for-analytics/create.sql

```diff
@@ -0,0 +1,118 @@
+-- Auto-infer the columns from the hits.parquet
+CREATE FOREIGN TABLE hits_auto () SERVER crunchy_lake_analytics OPTIONS (path 's3://clickhouse-public-datasets/hits_compatible/hits.parquet');
+
+-- To show representative behavior of Crunchy Bridge for Analytics, we point to the
+-- official hits.parquet file on S3 and load it to a local disk before running the queries
+SELECT * FROM crunchy_file_cache.add('s3://clickhouse-public-datasets/hits_compatible/hits.parquet');
```

Comment:
Ok, but need to split cold and hot results.
Cold results should not use the cache.

Reply:
Thanks for chiming in. We can change it that way if preferred, but the cache being referred to here is fully loading the data to disk rather than something like a buffer manager. 

Hence we included load time and felt it would be more accurate to have separate modes (query directly from S3 - no load time, query from local disk - with load time), where this PR adds the results of the latter.

Reply:
Hi @alexey-milovidov,

Have you had a chance to review Marco's comment above?

To clarify, the term `cache` here means that we download the files from S3 to the local machine's disk running the benchmark. That's why we use the title `"system": "Crunchy Bridge for Analytics (Parquet, local)"` to clearly describe our process.

In the future, we may add `"system": "Crunchy Bridge for Analytics (Parquet, remote)"`, where we query the Parquet file directly on S3.

Regarding your point:

> Cold results should not use the cache.

If you still believe we should avoid loading the data to the disk on the first run, we might need to adopt a more complex approach:

1. Disable auto-loading of the files to the disk (non-default behavior)
2. Execute the first run.
3. Load the files to the disk.
4. Execute the second and third runs.

Looking forward to your thoughts.

Reply:
Hi @onderkalaci 

As you mentioned, it is just a downloaded copy of s3 file on local disk, will an update query update both the local disk file as well as remote s3 file ?

Reply:
Reading through the Crunchy Bridge docs, I see two sections:

- Data Import & Export: https://docs.crunchybridge.com/analytics/data-import-and-export
- Caching: https://docs.crunchybridge.com/analytics/caching

May i ask what is the difference between the two.

Reply:
Hi @chhetripradeep 

> Hi @onderkalaci
> As you mentioned, it is just a downloaded copy of s3 file on local disk, will an update query update both the local > disk file as well as remote s3 file ?

As of the current release of Crunchy Bridge for Analytics, the system is read-only , you specify a file on `S3` and then query it.

> Reading through the Crunchy Bridge docs, I see two sections:
Data Import & Export: https://docs.crunchybridge.com/analytics/data-import-and-export


Crunchy Bridge for Analytics is an extension to Postgres. It means that in the same database cluster, you can have both regular Postgres tables (e.g., heap tables) as well as Crunchy Bridge for Analytics tables, the latter is provided via the extensions.

Just to be clear, in this benchmark, we are only running the tests against the Crunchy Bridge for Analytics tables.

Data Import piece is a capability for regular Postgres tables, not for analytics tables. It gives the ability to pull data into regular Postgres tables from files in s3 in formats like `parquet`, `json` or `csv`.


Data Export is the ability to push data from any Postgres data source (including regular Postgres tables as well as Crunchy Bridge for Analytics tables, or any query results etc) to `S3` in formats like `parquet`, `json` or `csv`.

So using data import feature and then running ClickBench would correspond to the vanilla postgresql benchmarks in https://github.com/ClickHouse/ClickBench/tree/main/postgresql , using the regular PostgreSQL storage and query engine.

> Caching: https://docs.crunchybridge.com/analytics/caching

Caching is for Crunchy Bridge for Analytics tables. As we described in the previous comments, `Crunchy Bridge for Analytics` tables are used to query files in `S3`. In many real use-cases, it is convenient to cache the data (e.g., download the files to the local machine) to prevent querying over the network repeatedly. This also happens automatically, but asynchronously in the background. Ongoing queries use (remote) range requests to `S3` until the data is available locally, so it is not like a buffer cache where the data is cached after the first query. It is effectively two modes of operation.


And, for this test, we use `"system": "Crunchy Bridge for Analytics (Parquet, local)"`  as the title such that we download files upfront and make them local. This is already the experience that many users have with the system.


Let us know if anything is missing.

Reply:
Thank you for the response @onderkalaci

As per [ClickBench Rules](https://github.com/ClickHouse/ClickBench?tab=readme-ov-file#rules-and-contribution), the first run of a query needs to be cold-run and second & third runs of the query can be hot-run, we would recommend to drop cache after every third run of a query. We need to ensure that we are following same rules for all databases otherwise results gets biased. So my understanding of the order of events is:

For each query:
- Execute the first run of the query. I guess It will auto-load the cache.
- Execute the second & third run of the query.
- Drop cache using `select crunchy_file_cache.remove('..')`

Reply:
Also i ran the scripts and got this [result](https://pastila.nl/?01782bfe/6b4ad825e86c30da0bfb02e1b2a34b3a#BaxmFv9DaOvsv5hiwTgBUg==) The result is bit different from the one from the PR i guess because of `psql timing` measure which is measuring client side latency. Would be good if you can add the location of client from which you expect to run the scripts from.

Reply:
hi @chhetripradeep,

Alright, let me re-run the tests, though the following is not 100% accurate.  The file download happens asynchronously, so without explicitly loading the file, the second run could still be "cold"

> For each query:
Execute the first run of the query. I guess It will auto-load the cache.
Execute the second & third run of the query.
Drop cache using select crunchy_file_cache.remove('..')

Also, "loading" / "dropping" the local file per query would make the running of the test quite long. So, instead we'd have something like this:
- Disable cache
- Run the first execution of all queries
- Enable cache & Synchronously load the data
- Run the second and the third executions of the queries

Are you OK with that?

Thanks!

Reply:
> Disable cache & run the first execution of all queries

If these steps make sure that first execution of all queries doesn't pick any data from cache, we are good to go with this approach.

Reply:
> > Disable cache & run the first execution of all queries
> 
> If these steps make sure that first execution of all queries doesn't pick any data from cache, we are good to go with this approach.

Yes, this makes sure the first execution of all queries use remote file.

Reply:
@chhetripradeep now pushed the following flow:

> Disable cache
Run the first execution of all queries
Enable cache & Synchronously load the data
Run the second and the third executions of the queries

Reply:
Hi @chhetripradeep,

Just to confirm, I'm assuming the next steps are with you. I'm waiting for any further feedback you might have.

Reply:
Hi @onderkalaci yes it's pending on me. Sorry for the delay. I will work on it in 1-2 days.

Reply:
The rules of ClickBench:

```
# Caching

If the system contains a cache for query results, it should be disabled.

It is okay if the system performs caching for source data (buffer pools and similar). If the cache or buffer pools can be flushed, they should be flushed before the first run of every query.

If the system contains a cache for intermediate data, that cache should be disabled if it is located near the end of the query execution pipeline, thus similar to a query result cache.
```

For me, it sounds like the rules were made before Cloud-native databases existed, i.e. they don't consider local caching of remote object store files. Anyways, either viewpoint is valid, I would say. We can go with the current approach (cold = remote access to S3, warm = read from locally downloaded files), but perhaps the README.md could point out more clearly what happens.

Reply:
>  but perhaps the README.md could point out more clearly what happens.

Sure, let me update the readme